### PR TITLE
impl(GCS+gRPC): `AsyncReaderConnection` with tracing

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #12798

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12799)

### CODE REVIEW

File: google/cloud/storage/internal/async/reader_connection_tracing.cc [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                                 {"message.starting_offset", payload.offset()},
+                             });
+      return r;
+    });
+  }
+
+ private:
+  opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span_;
+  std::unique_ptr<storage_experimental::AsyncReaderConnection> impl_;
+  std::int64_t count_ = 0;
```
Comment (devbww) [inline google/cloud/storage/internal/async/reader_connection_tracing.cc:84]: `#include <cstdint>`

**Review by devbww — APPROVED** [review]

File: google/cloud/storage/internal/async/reader_connection_tracing.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_ASYNC_READER_CONNECTION_TRACING_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_ASYNC_READER_CONNECTION_TRACING_H
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
```
Comment (dbolduc) [inline google/cloud/storage/internal/async/reader_connection_tracing.h:18]: I don't think you want these. I think you want to be able to call: `MakeTracingReaderConnection` from outside an `#if` guard.

Comment (coryan) [reply google/cloud/storage/internal/async/reader_connection_tracing.h:18]: It consumes a span... And the only thing that would call it is the tracing decorator for `AsyncConnection`.

File: google/cloud/storage/internal/async/reader_connection_tracing.cc [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+#include "google/cloud/storage/internal/async/reader_connection_tracing.h"
+#include "google/cloud/internal/opentelemetry.h"
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include <opentelemetry/trace/semantic_conventions.h>
+#endif  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include <sstream>
+#include <string>
+#include <thread>
+#include <utility>
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/async/reader_connection_tracing.cc]: Done.

File: google/cloud/storage/internal/async/reader_connection_tracing.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/async/reader_connection_tracing.cc:83]: Done.

File: google/cloud/storage/internal/async/reader_connection_tracing.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/async/reader_connection_tracing.h:18]: Fixed I think.

File: google/cloud/storage/internal/async/reader_connection_tracing.cc [hunk 4]
```diff
@@ -13,10 +13,11 @@
 // limitations under the License.
 
 #include "google/cloud/storage/internal/async/reader_connection_tracing.h"
```
Comment (dbolduc) [inline google/cloud/storage/internal/async/reader_connection_tracing.cc:15]: inconsequential comment: this include could go in the `#if` block too.

**Review by dbolduc — APPROVED** [review]

File: google/cloud/storage/internal/async/reader_connection_tracing.cc (hunk 4 above)
Comment (devbww) [reply google/cloud/storage/internal/async/reader_connection_tracing.cc:15]: I would say it is best to keep it as is, as the ability to make the inclusion is not a conditional thing, and should be exercised.  It is only the ability to do anything after the inclusion that is conditional.

**Pull request merged**